src/Model.php

```diff
@@ -404,40 +444,57 @@ public function validate($intent = null)
         return $errors;
     }
 
+    public function add($obj, $args = [])
+    {
+        $obj = $this->_add($obj, $args);
+        if ($obj instanceof Field) {
+            throw new Exception(['You should always use addField() for adding fields, not add()']);
+        }
+
+        return $obj;
+    }
+
     /**
      * Adds new field into model.
      *
-     * @param string $field
-     * @param array  $defaults
+     * @param string       $name
+     * @param array|object $seed
      *
      * @throws \atk4\core\Exception
      *
      * @return Field
      */
-    public function addField($field, $defaults = [])
+    public function addField($name, $seed = [])
     {
-        // compatibility: for field types
-        $class = $this->_default_seed_addField;
-        if (is_array($defaults) && isset($defaults['type'])) {
-            switch (strtolower($defaults['type'])) {
-                case 'boolean':
-                    $class = 'Boolean';
-                    break;
+        if (is_object($seed)) {
+            $field = $seed;
+        } else {
+            if ($this->persistence) {
+                $field = $this->persistence->fieldFactory($seed);
+            } else {
+                $field = $this->factory($this->mergeSeeds(
+                    $seed,
+                    isset($seed['type']) ? ($this->typeToFieldSeed[$seed['type']] ?? null) : null,
+                    Field::class
+                ), null, '\atk4\data\Field');
             }
         }
 
-        $field_object = $this->factory($this->mergeSeeds($defaults, $class), null, '\atk4\data\Field');
-        $this->add($field_object, $field);
-
-        return $field_object;
+        return $this->_addIntoCollection($name, $field, 'fields');
     }
 
+    protected $typeToFieldSeed = [
+        'boolean' => ['Boolean'],
+    ];
```

Comment:
I think this is good.

For other basic field types in future could be like:
```
'integer' => ['Number', 'unsigned'=>false, 'decimal'=>0, 'size'=>12],
'date' => ['DateTime', 'time'=>false, 'format'=>'Y-m-d'],
```
Basically it converts _basic_ field types to extended Field classes and sets some properties of these objects.